service/controller/resource/subnet/resource.go

```diff
@@ -0,0 +1,277 @@
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v4/client"
+	"github.com/giantswarm/azure-operator/v4/pkg/label"
+	"github.com/giantswarm/azure-operator/v4/service/controller/debugger"
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+	subnet "github.com/giantswarm/azure-operator/v4/service/controller/resource/subnet/template"
+)
+
+const (
+	credentialDefaultName = "credential-default"
+	credentialNamespace   = "giantswarm"
+	mainDeploymentName    = "subnet"
+	// Name is the identifier of the resource.
+	Name = "subnet"
+)
+
+type Config struct {
+	AzureClientsFactory *client.Factory
+	CtrlClient          ctrlclient.Client
+	Debugger            *debugger.Debugger
+	Logger              micrologger.Logger
+}
+
+type Resource struct {
+	azureClientsFactory *client.Factory
+	ctrlClient          ctrlclient.Client
+	debugger            *debugger.Debugger
+	logger              micrologger.Logger
+}
+
+type StorageAccountIpRule struct {
+	Value  string `json:"value"`
+	Action string `json:"action"`
+}
+
+func New(config Config) (*Resource, error) {
+	if config.AzureClientsFactory == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AzureClientsFactory must not be empty", config)
+	}
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+	if config.Debugger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Debugger must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &Resource{
+		azureClientsFactory: config.AzureClientsFactory,
+		ctrlClient:          config.CtrlClient,
+		debugger:            config.Debugger,
+		logger:              config.Logger,
+	}
+
+	return r, nil
+}
+
+// For every subnet declared in the `AzureCluster.Spec.NetworkSpec.Subnets` field, we submit a deployment to Azure to create the subnet.
+// The ipam handler is the one updating AzureCluster with the required subnets.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToAzureCluster(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	credentialSecret, err := r.getCredentialSecret(ctx, &cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	deploymentsClient, err := r.azureClientsFactory.GetDeploymentsClient(credentialSecret.Namespace, credentialSecret.Name)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	armTemplate, err := subnet.GetARMTemplate()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, allocatedSubnet := range cr.Spec.NetworkSpec.Subnets {
+		deploymentName := fmt.Sprintf("%s-%s", mainDeploymentName, allocatedSubnet.Name)
+		currentDeployment, err := deploymentsClient.Get(ctx, key.ClusterID(&cr), deploymentName)
+		if IsNotFound(err) {
+			// fallthrough
```

Comment:
Can we do this, since `currentDeployment` is used later in code? For example there is this few blocks later:

```
shouldSubmitDeployment := currentDeployment.IsHTTPStatus(404)
if !shouldSubmitDeployment {
	shouldSubmitDeployment, err = r.isDeploymentOutOfDate(ctx, cr, currentDeployment)
	if err != nil {
		return microerror.Mask(err)
	}
}
```

Reply:
if it doesn't exist `shouldSubmitDeployment` will be true. That code path always returns, so other occurrences of `currentDeployment` won't use a `nil` variable. Is that what you meant? Should I set `shouldSubmitDeployment` to true on line 102 instead?

Reply:
Sorry, I wasn't clear enough. I meant that if condition `IsNotFound(err)` is met, that would mean that `err != nil`, which would mean that returned `currentDeployment` was `nil`, so this line `currentDeployment.IsHTTPStatus(404)` would panic.

Reply:
That's if I got that right.

Reply:
We always get a `Response` back, even when the deployment is not found. So `currentDeployment` can't be nil.